fog/distribution/src/main.rs

```diff
@@ -355,23 +360,58 @@ fn worker_thread_entry(
 
         // Send to the next fog account
         let to_account = &fog_accounts[txs_created % fog_accounts.len()];
-
-        // Got our inputs, construct transaction.
-        let tx = build_tx(
-            &pending_spendable_txouts,
-            to_account,
-            &config,
-            &ledger_db,
-            fog_resolver.clone(),
-            &logger,
+        let env = Arc::new(
+            grpcio::EnvBuilder::new()
+                .name_prefix("FogPubkeyResolver-RPC".to_string())
+                .build(),
         );
 
-        txs_created += 1;
+        let conn = GrpcFogReportConnection::new(env, logger.clone());
+        let fog_uri = FogUri::from_str(
+            fog_accounts[0]
+                .default_subaddress()
+                .fog_report_url()
+                .expect("No fog report url"),
+        )
+        .expect("Could not parse fog url");
+        // Sometime transactions could not be submitted before tombstone block passed,
+        // so loop until transactions can be submmitted
+        loop {
+            // Got our inputs, construct transaction.
+            let tx = build_tx(
+                &pending_spendable_txouts,
+                to_account,
+                &config,
+                &ledger_db,
+                fog_resolver.clone(),
+                &logger,
+            );
 
-        // Submit tx
-        if submit_tx(txs_created, &conns, &tx, &config, &logger) {
-            let mut map = TX_PUB_KEY_TO_ACCOUNT_KEY.lock().unwrap();
-            map.insert(tx.prefix.outputs[0].public_key, to_account.clone());
+            // Submit tx
+            if submit_tx(txs_created, &conns, &tx, &config, &logger) {
+                txs_created += 1;
+                let mut map = TX_PUB_KEY_TO_ACCOUNT_KEY.lock().unwrap();
+                map.insert(tx.prefix.outputs[0].public_key, to_account.clone());
+                break;
+            } else {
+                //each worker thread should build its own FogResolver,
+                //if submit fails, it should trash and rebuild the FogResolver to ensure it has
+                // fresh fog
+                let responses = rebuild_fogresolver(&fog_uri, &conn, &logger);
+
+                let report_verifier = {
+                    let mr_signer_verifier =
+                        fog_ingest_enclave_measurement::get_mr_signer_verifier(None);
+
+                    let mut verifier = Verifier::default();
+                    verifier.debug(DEBUG_ENCLAVE).mr_signer(mr_signer_verifier);
+                    verifier
+                };
+
+                fog_resolver = FogResolver::new(responses.clone(), &report_verifier)
+                    .expect("Could not get FogResolver");
```

Comment:
This looks really good

I think it is slightly better if the report_verifier and Fog resolver::new calls are also in the function you are adding, because then that code won't appear in two places